### PR TITLE
Tasks: Track token threshold config

### PR DESCRIPTION
**Pull request by PatricioHenderson**
Not sure if `defaultTokenThreshold` values should be inside `Task entity` , or as It's been made, as an entire separate entity with a relation  to the Task

### CODE REVIEW

**Comment by facuspagnuolo** [comment]
> Not sure if `defaultTokenThreshold` values should be inside `Task entity` , or as It's been made, as an entire separate entity with a relation  to the Task

I like the idea of having a separate entity for it. Actually, I'd seize that information to normalize more the custom settings values. See my comment about that on PR #3. I think we should achieve the same thing here.

File: schema.graphql [hunk 1]
```diff
@@ -86,8 +86,28 @@ type Task @entity {
...
+type CustomTokenThreshold @entity {
+  id: ID!
+  task: Task!
+  token: ERC20!
+  thresholdToken: ERC20!
+  tokenThreshold: TokenThreshold!
+}
+
+
+
```
Comment (facuspagnuolo) [inline schema.graphql:110]:
```suggestion

```

File: src/Task.ts [hunk 2]
```diff
@@ -22,6 +25,45 @@ export function handleBalanceConnectorsSet(event: BalanceConnectorsSet): void {
...
+
+  const customTokenThresholdId = getTokenThresholdId(task, event.params.token)
+  const tokenThreshold = new TokenThreshold(customTokenThresholdId)
+
+  tokenThreshold.task = task.id
+  tokenThreshold.token = loadOrCreateERC20(event.params.token).id
+  tokenThreshold.min = event.params.min
+  tokenThreshold.max = event.params.max
+
+  tokenThreshold.save()
```
Comment (facuspagnuolo) [inline src/Task.ts:40]:
```suggestion
  tokenThreshold.save()
```

File: src/Registry.ts [hunk 3]
```diff
@@ -31,3 +31,19 @@ export function loadOrCreateImplementation(address: Address): Implementation {
...
+    tokenThreshold = new TokenThreshold(tokenThresholdId)
+    tokenThreshold.task = task.id
+    tokenThreshold.token = tokenAddress.toHexString()
+    tokenThreshold.min = BigInt.zero()
+    tokenThreshold.max = BigInt.zero()
+    tokenThreshold.save()
+  }
+
+  return tokenThreshold.id
+}
```
Comment (facuspagnuolo) [inline src/Registry.ts:49]: Let's place this function inside of `Task.ts`

File: src/Task.ts [hunk 4]
```diff
@@ -22,6 +25,45 @@ export function handleBalanceConnectorsSet(event: BalanceConnectorsSet): void {
...
 }
 
+export function handleCustomTokenThresholdSet(event: CustomTokenThresholdSet): void {
+  const task = Task.load(event.address.toHexString())
+  if (task == null) return log.warning('Missing task entity {}', [event.address.toHexString()])
+
+  const customTokenThresholdId = getTokenThresholdId(task, event.params.token)
+  const tokenThreshold = new TokenThreshold(customTokenThresholdId)
+
+  tokenThreshold.task = task.id
```
Comment (facuspagnuolo) [inline src/Task.ts:35]:
```suggestion
  tokenThreshold.task = task.id
```

File: src/Task.ts [hunk 5]
```diff
@@ -22,6 +25,45 @@ export function handleBalanceConnectorsSet(event: BalanceConnectorsSet): void {
   task.save()
 }
 
+export function handleCustomTokenThresholdSet(event: CustomTokenThresholdSet): void {
+  const task = Task.load(event.address.toHexString())
+  if (task == null) return log.warning('Missing task entity {}', [event.address.toHexString()])
+
+  const customTokenThresholdId = getTokenThresholdId(task, event.params.token)
+  const tokenThreshold = new TokenThreshold(customTokenThresholdId)
```
Comment (facuspagnuolo) [inline src/Task.ts:33]: We should use `loadOrCreate` here

File: src/Task.ts [hunk 6]
```diff
@@ -22,6 +25,45 @@ export function handleBalanceConnectorsSet(event: BalanceConnectorsSet): void {
...
+
+  tokenThreshold.task = task.id
+  tokenThreshold.token = loadOrCreateERC20(event.params.token).id
+  tokenThreshold.min = event.params.min
+  tokenThreshold.max = event.params.max
+
+  tokenThreshold.save()
+
+  const customTokenThreshold = new CustomTokenThreshold(customTokenThresholdId)
+
```
Comment (facuspagnuolo) [inline src/Task.ts:43]:
```suggestion
  const customTokenThreshold = new CustomTokenThreshold(customTokenThresholdId)
```

File: src/Task.ts [hunk 7]
```diff
@@ -22,6 +25,45 @@ export function handleBalanceConnectorsSet(event: BalanceConnectorsSet): void {
...
+  customTokenThreshold.task = task.id
+  customTokenThreshold.token = loadOrCreateERC20(event.params.token).id
+  customTokenThreshold.thresholdToken = loadOrCreateERC20(event.params.thresholdToken).id
+  customTokenThreshold.tokenThreshold = tokenThreshold.id
+
+  customTokenThreshold.save()
```
Comment (facuspagnuolo) [inline src/Task.ts:49]:
```suggestion
  customTokenThreshold.save()
```

File: schema.graphql [hunk 8]
```diff
@@ -86,8 +86,28 @@ type Task @entity {
...
+  min: BigInt!
+  max: BigInt!
+}
+
+type CustomTokenThreshold @entity {
+  id: ID!
+  task: Task!
+  token: ERC20!
+  thresholdToken: ERC20!
+  tokenThreshold: TokenThreshold!
```
Comment (facuspagnuolo) [inline schema.graphql:106]: I think `thresholdToken` is actually in `TokenThreshold`, right?

```suggestion
  threshold: TokenThreshold!
```

File: src/Deployer.ts [hunk 9]
```diff
@@ -83,6 +83,7 @@ export function handleTaskDeployed(event: TaskDeployed): void {
   task.timeLockDelay = BigInt.zero()
   task.timeLockExecutionPeriod = BigInt.zero()
   task.timeLockExpiration = BigInt.zero()
+  task.defaultTokenThreshold = loadOrCreateDefaultTokenThreshold(task, event.params.instance).toString()
```
Comment (facuspagnuolo) [inline src/Deployer.ts:86]: Why passing `event.params.instance` here?

File: src/Registry.ts [hunk 10]
```diff
@@ -31,3 +31,19 @@ export function loadOrCreateImplementation(address: Address): Implementation {
...
 }
+
+export function loadOrCreateDefaultTokenThreshold(task: Task, tokenAddress: Address): string {
+  const tokenThresholdId = task.id
+  let tokenThreshold = TokenThreshold.load(tokenThresholdId)
+
+  if (tokenThreshold === null) {
+    tokenThreshold = new TokenThreshold(tokenThresholdId)
+    tokenThreshold.task = task.id
+    tokenThreshold.token = tokenAddress.toHexString()
```
Comment (facuspagnuolo) [inline src/Registry.ts:42]: I think there is a misconception here, if there is no threshold then there is no token defined for it. Happy to discuss this further if needed.

File: src/Task.ts [hunk 11]
```diff
@@ -22,6 +25,45 @@ export function handleBalanceConnectorsSet(event: BalanceConnectorsSet): void {
...
+}
+
+export function handleDefaultTokenThresholdSet(event: DefaultTokenThresholdSet): void {
+  const task = Task.load(event.address.toHexString())
+  if (task == null) return log.warning('Missing task entity {}', [event.address.toHexString()])
+
+  const defaultTokenThresholdId = task.id
+
```
Comment (facuspagnuolo) [inline src/Task.ts:57]:
```suggestion
  const defaultTokenThresholdId = task.id
```

File: src/Task.ts [hunk 12]
```diff
@@ -22,6 +25,45 @@ export function handleBalanceConnectorsSet(event: BalanceConnectorsSet): void {
...
+  const defaultTokenThreshold = new TokenThreshold(defaultTokenThresholdId)
+  defaultTokenThreshold.task = task.id
+  defaultTokenThreshold.token = loadOrCreateERC20(event.params.token).id
+  defaultTokenThreshold.min = event.params.min
+  defaultTokenThreshold.max = event.params.max
+
+  defaultTokenThreshold.save()
```
Comment (facuspagnuolo) [inline src/Task.ts:64]:
```suggestion
  defaultTokenThreshold.save()
```

File: src/Task.ts [hunk 13]
```diff
@@ -22,6 +25,45 @@ export function handleBalanceConnectorsSet(event: BalanceConnectorsSet): void {
...
+  customTokenThreshold.save()
+}
+
+export function handleDefaultTokenThresholdSet(event: DefaultTokenThresholdSet): void {
+  const task = Task.load(event.address.toHexString())
+  if (task == null) return log.warning('Missing task entity {}', [event.address.toHexString()])
+
+  const defaultTokenThresholdId = task.id
+
+  const defaultTokenThreshold = new TokenThreshold(defaultTokenThresholdId)
```
Comment (facuspagnuolo) [inline src/Task.ts:58]: We should use `loadOrCreate` here

File: src/Task.ts [hunk 14]
```diff
@@ -113,3 +155,7 @@ export function getExecutionType(address: Address): Bytes {
   log.warning('EXECUTION_TYPE() call reverted for task {}', [address.toHexString()])
   return Bytes.fromUTF8('')
 }
+
+export function getTokenThresholdId(task: Task, token: Address): string {
```
Comment (facuspagnuolo) [inline src/Task.ts:159]: This is `getCustomTokenThresholdId`, isn't it?

File: schema.graphql [hunk 15]
```diff
@@ -86,6 +86,8 @@ type Task @entity {
   timeLockDelay: BigInt!
   timeLockExecutionPeriod: BigInt!
   timeLockExpiration: BigInt!
+  defaultTokenThreshold: TokenThreshold!
```
Comment (facuspagnuolo) [inline schema.graphql:89]: Let's apply the same change I did in #11, I know I suggested to have a non-null value here, but it's failing and actually makes sense to make it nullable.

File: schema.graphql [hunk 16]
```diff
@@ -105,6 +107,20 @@ type CustomVolumeLimit @entity {
...
+  min: BigInt!
+  max: BigInt!
+}
+
+type CustomTokenThreshold @entity {
+  id: ID!
+  task: Task!
+  token: ERC20!
+  threshold: TokenThreshold!
+}
```
Comment (facuspagnuolo) [inline schema.graphql:123]:
```suggestion
}

```

File: schema.graphql [hunk 17]
```diff
@@ -105,6 +107,20 @@ type CustomVolumeLimit @entity {
   volumeLimit: VolumeLimit!
 }
 
+type TokenThreshold @entity {
+  id: ID!
+  task: Task!
```
Comment (facuspagnuolo) [inline schema.graphql:112]: I'd remove the task from here

File: schema.graphql (hunk 15 above)
Comment (PatricioHenderson) [reply schema.graphql]: Sure, no problem!

**Review by facuspagnuolo — APPROVED** [review]
LGTM!

**Pull request merged**